core/src/main/java/org/apache/calcite/plan/volcano/RelSubset.java

```diff
@@ -144,6 +141,22 @@ private void computeBestCost(RelOptPlanner planner) {
     }
   }
 
+  public void setDerived() {
+    state |= DERIVED;
+  }
+
+  public void setRequired() {
+    state |= REQUIRED;
+  }
+
+  public boolean isDerived() {
+    return (state & DERIVED) == DERIVED;
+  }
+
+  public boolean isRequired() {
```

Comment:
I am confused here. Is RelSubSet' state either DERIVED or REQUIRED?

Reply:
The traitset can be both required and derived.

Reply:
Got it.

Reply:
Maybe we can figure out a way that makes it more clear.